### PR TITLE
[package upgrades] change upgrade criteria for struct type params

### PR DESCRIPTION
**Pull request by tzakian**

### CODE REVIEW

File: crates/sui-adapter-transactional-tests/tests/upgrade/remove_phantom.move [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+}
+
+//# upgrade --package A0 --upgrade-capability 1,1 --sender A
+module A1::base {
+    struct Foo<T> { }
+}
+
+//# upgrade --package A0 --upgrade-capability 1,1 --sender A
+module A1::base {
+    struct Foo<T: store> { }
```
Comment (amnn) [inline crates/sui-adapter-transactional-tests/tests/upgrade/remove_phantom.move:18]: You are also adding a `store` here, is that intentionaly?  That would be incompatible anyway but for another reason.

File: external-crates/move/move-binary-format/src/compatibility.rs [hunk 2]
```diff
@@ -247,39 +253,56 @@ fn fun_type_parameters_compatibile(
...
 ) -> bool {
     old_type_parameters.len() == new_type_parameters.len()
         && old_type_parameters.iter().zip(new_type_parameters).all(
             |(old_type_parameter, new_type_parameter)| {
-                type_parameter_phantom_decl_compatibile(old_type_parameter, new_type_parameter)
-                    && type_parameter_constraints_compatibile(
-                        old_type_parameter.constraints,
-                        new_type_parameter.constraints,
-                    )
+                type_parameter_phantom_decl_compatibile(
```
Comment (amnn) [inline external-crates/move/move-binary-format/src/compatibility.rs:263]: nit, spelling on `compatibile`.

File: external-crates/move/move-binary-format/src/compatibility.rs [hunk 3]
```diff
@@ -247,39 +253,56 @@ fn fun_type_parameters_compatibile(
...
+                    disallow_changing_generic_abilities,
+                    old_type_parameter,
+                    new_type_parameter,
+                ) && type_parameter_constraints_compatibile(
```
Comment (amnn) [inline external-crates/move/move-binary-format/src/compatibility.rs:267]: ditto

File: external-crates/move/move-binary-format/src/compatibility.rs (hunk 2 above)
Comment (amnn) [reply external-crates/move/move-binary-format/src/compatibility.rs:263]: But also recognise that this wasn't introduced in this PR.

File: external-crates/move/move-binary-format/src/compatibility.rs [hunk 4]
```diff
@@ -247,39 +253,56 @@ fn fun_type_parameters_compatibile(
...
+    disallow_changing_generic_abilities: bool,
     old_type_parameter: &StructTypeParameter,
     new_type_parameter: &StructTypeParameter,
 ) -> bool {
-    // old_type_paramter.is_phantom => new_type_parameter.is_phantom
-    !old_type_parameter.is_phantom || new_type_parameter.is_phantom
+    if disallow_changing_generic_abilities {
+        // phantom/non-phantom cannot change from one version to the next.
+        old_type_parameter.is_phantom && new_type_parameter.is_phantom
+            || !old_type_parameter.is_phantom && !new_type_parameter.is_phantom
```
Comment (amnn) [inline external-crates/move/move-binary-format/src/compatibility.rs:301]: Just use `==`?

```suggestion
        // phantom/non-phantom cannot change from one version to the next.
        old_type_parameter.is_phantom == new_type_parameter.is_phantom
```

**Review by amnn — APPROVED** [review]
LGTM!  Couple nits, thanks @tzakian .

File: crates/sui-adapter-transactional-tests/tests/upgrade/remove_phantom.move (hunk 1 above)
Comment (tzakian) [reply crates/sui-adapter-transactional-tests/tests/upgrade/remove_phantom.move:18]: Yep! But I just wanted to make sure we were covering that as well here.

File: external-crates/move/move-binary-format/src/compatibility.rs (hunk 4 above)
Comment (tzakian) [reply external-crates/move/move-binary-format/src/compatibility.rs:301]: Magic.... 🤦🏻

**Pull request merged**